components/ai_chat/resources/page/components/filter_menu/tools_menu.tsx

```diff
@@ -4,43 +4,113 @@
 // You can obtain one at https://mozilla.org/MPL/2.0/.
 
 import * as React from 'react'
+import Button from '@brave/leo/react/button'
+import Icon from '@brave/leo/react/icon'
 
 import {
   ActionEntry,
   ActionType,
+  SmartMode,
 } from 'components/ai_chat/resources/common/mojom'
 import FilterMenu, { Props } from './filter_menu'
 import { matches } from './query'
 import styles from './style.module.scss'
+import { getLocale } from '$web-common/locale'
+import { useAIChat } from '../../state/ai_chat_context'
 
 type ToolsMenuProps = {
   handleClick: (type: ActionType) => void
-} & Pick<Props<ActionEntry>, 'categories' | 'isOpen' | 'setIsOpen' | 'query'>
+  handleSmartModeClick?: (smartMode: SmartMode) => void
+  handleSmartModeEdit?: (smartMode: SmartMode) => void
+} & Pick<
+  Props<ExtendedActionEntry>,
+  'categories' | 'isOpen' | 'setIsOpen' | 'query'
+>
 
-function matchesQuery(query: string, entry: ActionEntry) {
-  return entry.details && matches(query, entry.details.label)
+// Extended entry type that can include smart modes before we migrate
+// existing slash tools (actions) to smart modes.
+type ExtendedActionEntry = ActionEntry & {
+  smartMode?: SmartMode
+}
+
+function matchesQuery(query: string, entry: ExtendedActionEntry) {
+  if (entry.details) {
+    return matches(query, entry.details.label)
+  } else if (entry.smartMode) {
+    return matches(query, entry.smartMode.shortcut)
+  }
+  return false
 }
 
 export default function ToolsMenu(props: ToolsMenuProps) {
+  const { smartModes } = useAIChat()
+
+  // Create combined categories including smart modes
+  const combinedCategories = React.useMemo(() => {
+    const categories = []
+
+    if (
+      smartModes.length > 0
+      && props.handleSmartModeClick
+      && props.handleSmartModeEdit
+    ) {
+      const smartModeEntries: ExtendedActionEntry[] = smartModes.map(
+        (mode: SmartMode) => ({
+          smartMode: mode,
+          subheading: undefined,
+          details: undefined,
+        }),
+      )
+
+      categories.push({
+        category: getLocale(S.CHAT_UI_SMART_MODES_GROUP),
+        entries: smartModeEntries,
+      })
+    }
+
+    categories.push(...props.categories)
+
+    return categories
+  }, [smartModes])
+
   return (
     <FilterMenu
-      categories={props.categories}
+      categories={combinedCategories}
       isOpen={props.isOpen}
       setIsOpen={props.setIsOpen}
       query={props.query}
       matchesQuery={matchesQuery}
     >
-      {(item) =>
-        !item.details ? (
+      {(item: ExtendedActionEntry) =>
```

Comment:
I think we shouldn't need this type here - possibly the type of `combinedCategories` isn't being inferred properly.

If that's the case, we should probably add a type param to the `useMemo` above